src/nbl/asset/utils/CGeometryCreator.cpp

```diff
@@ -423,239 +395,373 @@ core::smart_refctd_ptr<ICPUPolygonGeometry> CGeometryCreator::createSphere(float
+		u1 = hlsl::normalize(hlsl::cross(v0ToApex, u1)) * angleWeight;
 
 		uint32_t prevVertexIndex = i == 0 ? (tesselation - 1) : i - 1;
-		core::vectorSIMDf u2 = core::vectorSIMDf(baseVertices[prevVertexIndex].pos[0], baseVertices[prevVertexIndex].pos[1], baseVertices[prevVertexIndex].pos[2]);
-		u2 -= core::vectorSIMDf(baseVertices[i].pos[0], baseVertices[i].pos[1], baseVertices[i].pos[2]);
-		angleWeight = std::acos(core::dot(core::normalize(apexVertexCoords), core::normalize(u2)).x);
-		u2 = core::normalize(core::cross(u2, v0ToApex)) * angleWeight;
+		hlsl::float32_t3 u2 = hlsl::float32_t3(positions[prevVertexIndex].x, positions[prevVertexIndex].y, positions[prevVertexIndex].z);
+		u2 -= simdPosition;
+		angleWeight = std::acos(hlsl::dot(hlsl::normalize(apexVertexCoords), hlsl::normalize(u2)));
+		u2 = hlsl::normalize(hlsl::cross(u2, v0ToApex)) * angleWeight;
+
+
+		const auto baseNormal = quantNormalCache->quantize<NormalCacheFormat>(hlsl::normalize(u1 + u2));
+		memcpy(normals + i, &baseNormal, sizeof(baseNormal));
 
-		baseVertices[i].normal = quantNormalCache->quantize<EF_A2B10G10R10_SNORM_PACK32>(core::normalize(u1 + u2));
-		apexVertices[i].normal = quantNormalCache->quantize<EF_A2B10G10R10_SNORM_PACK32>(core::normalize(u1));
+		const auto apexNormal = quantNormalCache->quantize<NormalCacheFormat>(hlsl::normalize(u1));
+		memcpy(normals + apexVertexBase_i + i, &apexNormal, sizeof(apexNormal));
 	}
 
-	auto idxBuf = asset::ICPUBuffer::create({ 3u * tesselation * sizeof(uint16_t) });
-	uint16_t* indices = (uint16_t*)idxBuf->getPointer();
+	CPolygonGeometryManipulator::recomputeContentHashes(retval.get());
+	return retval;
+}
 
-	const uint32_t firstIndexOfBaseVertices = 0;
-	const uint32_t firstIndexOfApexVertices = tesselation;
-	for (uint32_t i = 0; i < tesselation; i++)
+core::vector<core::smart_refctd_ptr<ICPUPolygonGeometry>> CGeometryCreator::createArrow(
+	const uint32_t tesselationCylinder,
+	const uint32_t tesselationCone,
+	const float height,
+	const float cylinderHeight,
+	const float width0,
+	const float width1
+) const
+{
+	assert(height > cylinderHeight);
+
+	using position_t = hlsl::float32_t3;
+
+	auto cylinder = createCylinder(width0, cylinderHeight, tesselationCylinder);
+	auto cone = createCone(width1, height-cylinderHeight, tesselationCone);
+
+	auto conePositions = reinterpret_cast<position_t*>(cone->getPositionView().src.buffer->getPointer());
+
+	const auto coneVertexCount = cone->getPositionView().getElementCount();
+	
+	for (auto i = 0ull; i < coneVertexCount; ++i)
 	{
-		indices[i * 3] = firstIndexOfApexVertices + i;
-		indices[(i * 3) + 1] = firstIndexOfBaseVertices + i;
-		indices[(i * 3) + 2] = i == (tesselation - 1) ? firstIndexOfBaseVertices : firstIndexOfBaseVertices + i + 1;
+		auto& conePosition = conePositions[i];
+		core::vector3df_SIMD newPos(conePosition.x, conePosition.y, conePosition.z);
+		newPos.rotateYZByRAD(-1.5707963268);
+
+		conePosition = {newPos.x, newPos.y, newPos.z};
 	}
```

Comment:
no, a geometry collection has transform matrices for each geometry, you shouldn't need to transform vertices in the actual vertex buffer